### PR TITLE
[HIP] clean up Teams

### PR DESCRIPTION
**Pull request by Rombur**
This PR removes some useless macro guards. We usually don't guard device code in HIP because only the hip compiler should have access to this code. This PR also fixes a couple of bugs where we assumed that the warp size was 32.

This PR might conflict with #3215 depends how smart git is :smiley:

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP_Instance.hpp [hunk 1]
```diff
@@ -55,6 +55,7 @@ namespace Impl {
 
 struct HIPTraits {
   static int constexpr WarpSize       = 64;
+  static int constexpr WarpIndexMask  = 0x003f;
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Instance.hpp:58]: Comment it is the hexadecimal for 63 for those like me who would need to look it up.

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp [hunk 2]
```diff
@@ -197,8 +197,9 @@ class TeamPolicyInternal<Kokkos::Experimental::HIP, Properties...>
 
     // Allow only power-of-two vector_length
     if (!(is_integral_power_of_two(test_vector_length))) {
-      int test_pow2 = 1;
-      for (int i = 0; i < 5; i++) {
+      int test_pow2           = 1;
+      int constexpr warp_size = Experimental::Impl::HIPTraits::WarpSize;
+      while (test_pow2 < warp_size) {
         test_pow2 = test_pow2 << 1;
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Parallel_Team.hpp:203]:
```suggestion
        test_pow2 <<= 1;
```

File: core/src/HIP/Kokkos_HIP_Team.hpp [hunk 3]
```diff
@@ -366,20 +309,21 @@ class HIPTeamMember {
...
-    const int wx = tid & HIPTraits::WarpIndexMask;
-    const int wy = tid >> HIPTraits::WarpIndexShift;
+    const int wn = (nid + Experimental::Impl::HIPTraits::WarpIndexMask) >>
+                   Experimental::Impl::HIPTraits::WarpIndexShift;
+    const int wx = tid & Experimental::Impl::HIPTraits::WarpIndexMask;
+    const int wy = tid >> Experimental::Impl::HIPTraits::WarpIndexShift;
 
     //------------------------
-    {  // Intra warp shuffle reduction from contributing CUDA threads
+    {  // Intra warp shuffle reduction from contributing HIP threads
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Team.hpp:318]: :)
```
git grep -i cuda core/src/HIP/
core/src/HIP/Kokkos_HIP_ReduceScan.hpp:    // problem was also found in the CUDA backend with CUDA clang
core/src/HIP/Kokkos_HIP_Team.hpp:    {  // Intra warp shuffle reduction from contributing CUDA threads
core/src/HIP/Kokkos_HIP_Team.hpp:#endif /* #ifndef KOKKOS_CUDA_TEAM_HPP */

File: core/src/HIP/Kokkos_HIP_Team.hpp [hunk 4]
```diff
@@ -402,7 +402,7 @@ class HIPTeamMember {
       if (0 == wy) {
         // Start fan-in at power of two covering nentry
 
-        for (int i = (1 << (32 - __clz(nentry - 1))); (i >>= 1);) {
+        for (int i = (1 << (warp_size - __clz(nentry - 1))); (i >>= 1);) {
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Team.hpp:369]: I think that was the only one.
```
git grep "\<32" core/src/HIP/
core/src/HIP/Kokkos_HIP_Instance.cpp:    // theoretically, we can get 40 WF's / CU, but only can sustain 32
core/src/HIP/Kokkos_HIP_Instance.cpp:    m_maxBlocksPerSM = 32;
core/src/HIP/Kokkos_HIP_Instance.cpp:    m_regsPerSM          = 262144 / 32;
core/src/HIP/Kokkos_HIP_Instance.hpp:  static int constexpr ConstantMemoryUsage        = 0x008000; /* 32k bytes */
core/src/HIP/Kokkos_HIP_Team.hpp:        for (int i = (1 << (32 - __clz(nentry - 1))); (i >>= 1);) {

File: core/src/HIP/Kokkos_HIP_Team.hpp (hunk 3 above)
Comment (dalg24) [reply core/src/HIP/Kokkos_HIP_Team.hpp:318]: Maybe fix the `KOKKOS_CUDA_TEAM_HPP` too

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp [hunk 5]
```diff
@@ -200,7 +200,7 @@ class TeamPolicyInternal<Kokkos::Experimental::HIP, Properties...>
       int test_pow2           = 1;
       int constexpr warp_size = Experimental::Impl::HIPTraits::WarpSize;
       while (test_pow2 < warp_size) {
-        test_pow2 = test_pow2 << 1;
+        test_pow2 << 1;
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Parallel_Team.hpp:203]:
```suggestion
        test_pow2 <<= 1;
```

**Comment by masterleinad** [comment]
The indentation is off.

**Pull request merged**